### PR TITLE
mix deps.add

### PR DESCRIPTION
**Pull request by TheFirstAvenger**
PR to continue discussion on https://github.com/elixir-lang/elixir/pull/10291

### CODE REVIEW

File: lib/mix/lib/mix/dep/adder.ex [hunk 1]
```diff
@@ -0,0 +1,132 @@
...
+      |> String.starts_with?("{:" <> app)
+    end)
+    |> case do
+      nil ->
+        :ok
+
+      found ->
+        Mix.raise("#{app} already exists in mix.exs as #{String.trim(found)}")
+    end
+  end
```
Comment (Eiji7) [inline lib/mix/lib/mix/dep/adder.ex:83]: `String`-based checks are unsafe. Instead it should be done on parsing `AST`. Simplest example which would not `raise`:
```elixir
def project do
  [
    # …
    deps: [{:my_dep, …}],
    # …
  ]
end
```

File: lib/mix/lib/mix/dep/adder.ex [hunk 2]
```diff
@@ -0,0 +1,132 @@
...
+        Enum.reverse(acc) ++
+          [line1, indent("[", line1), indent(dep_line, line2), indent("]", line1) | rest]
+
+      String.trim(line1) == "defp deps do" && String.starts_with?(String.trim(line2), "[{:") ->
+        Enum.reverse(acc) ++
+          [line1, String.replace(line2, ~r/\[/, "[#{dep_line}, ", global: false) | rest]
+
+      true ->
+        add_dep_line([line2 | rest], [line1 | acc], dep_line)
+    end
```
Comment (Eiji7) [inline lib/mix/lib/mix/dep/adder.ex:112]: Same as `check_for_app!/2` … When parsing `AST` we few options:
1. Invalid `mix.exs` file (typo, missing file etc.)
2. if `:deps` option is `variable` or raw `List` - if so do: `[{:my_dep, …} | variable]` or simply insert `:my_dep` as first element of `List`
3. if `:deps` option call 0-arity function - in such case for last expression try point 2
4. otherwise display proper message

Also in current code such example would fail:
```elixir
defp deps, do: […]
```

File: lib/mix/lib/mix/dep/adder.ex [hunk 3]
```diff
@@ -0,0 +1,132 @@
...
+  end
+
+  defp add_dep_line(_, _, _), do: :deps_not_found
+
+  defp indent(line, parent_line) do
+    String.duplicate(" ", indented(parent_line) + 2) <> line
```
Comment (Eiji7) [inline lib/mix/lib/mix/dep/adder.ex:118]: `2` spaces are just standard however trust me or not there are some developers which ignores `Elixir` standards, because for example in other languages they write code differently - I really can't get it, but this is only small :+1:  for working on `AST`

**Review by Eiji7 — COMMENTED** [review]
TLDR:
This should be done using traversing of `AST`. When working with `String` it may have many cases which would unexpectedly fail.

**Comment by TheFirstAvenger** [comment]
@Eiji7 Yes, no matter what solution is proposed, there will always be a way to break it. For example, if your deps function was multi-headed and took a parameter such as `Mix.env`, or if it called out to a different exs file, or any one of a number of other modifications you could make to it. The idea here is that we have a solution that will work for 100% of projects immediately after running `mix new`, and will work on any project where there isn't structural changes to `defp deps do`. For the other cases where the developer has changed the structure, the current code recognize the fact that we couldn't identify the correct modification, and simply say "here is the line you need to add, do it manually". Given that some package tools _only_ give you that manual option, it doesn't seem like a reason not to add the feature.

As mentioned in the other PR, this first step allows us to make the feature available for the majority of cases, and then subsequently work on incrementally improve the detection, either tweaking the string parsing or when Wojtek has an AST formatter based solution.

In my opinion, waiting for the perfect solution that can detect and automatically add dependencies in every single possible situation unnecessarily delays this useful feature which we know works when the `mix new` standards are followed, and which is still useful when they are not.

File: lib/mix/lib/mix/dep/adder.ex (hunk 1 above)
Comment (TheFirstAvenger) [reply lib/mix/lib/mix/dep/adder.ex:83]: This would result in the following output in the current PR:
```
Could not find `defp deps do` in your mix.exs. Please add this line to mix.exs manually:
{:foo, ">= 0.0.0"}
```

File: lib/mix/lib/mix/dep/adder.ex (hunk 2 above)
Comment (TheFirstAvenger) [reply lib/mix/lib/mix/dep/adder.ex:112]: Similar to above, this would currently result in:
```
Could not find `defp deps do` in your mix.exs. Please add this line to mix.exs manually:
{:foo, ">= 0.0.0"}
```
although tweaking the code to be able to detect that one-liner would be pretty trivial.

File: lib/mix/lib/mix/dep/adder.ex (hunk 3 above)
Comment (TheFirstAvenger) [reply lib/mix/lib/mix/dep/adder.ex:118]: We currently check that the code is formatted, and then format after modifying it, so this would not trip up the system. The end user would get:
```
mix_exs was not formatted. Please run `mix format` and retry, or add this line to mix.exs manually:
{:foo, ">= 0.0.0"}
```

**Comment by Eiji7** [comment]
> @TheFirstAvenger In my opinion, waiting for the perfect solution that can detect and automatically add dependencies in every single possible situation unnecessarily delays this useful feature which we know works when the mix new standards are followed, and which is still useful when they are not.

So it would work only for `mix new`. Some developers have their own templates (simplest example: without extra comments) either copy-pasted manually, via `mix` archive or using some IDE template feature.

> if your deps function was multi-headed and took a parameter such as Mix.env, or if it called out to a different exs file, or any one of a number of other modifications you could make to it

Not really … in theory such cases are complicated, but in practice almost nobody with so complex modifications would like to use such a task.

I fully understand that we should not wait for perfect solution, but there are always 2 edges and adding such feature as fast as possible is another one. What I pointed are simplest cases for raw `List` (or `List` variable) and 0-arity call. I believe that `AST`-based solution would handle all edge-cases of custom templates and is exactly in the middle between "perfect" and simple code.

Also I see that it's intended mainly for `mix new`, but in such way we would have many similar solutions for different templates which simply does not makes sense. I understand that it could be extended, but simply following "as fast as possible" it's faster to fork and modify this task (for only one edge-case) rather than wait for more general-purpose solution.

That's my 2 cents :slightly_smiling_face:

**Comment by TheFirstAvenger** [comment]
> Some developers have their own templates (simplest example: without extra comments) either copy-pasted manually, via mix archive or using some IDE template feature.

Yes, and if they wanted to use this feature, as long as they still had a `defp deps do` function, auto modification would work. If they didn't, the feature would give them the exact code they need to add manually.

> Not really … in theory such cases are complicated, but in practice almost nobody with so complex modifications would like to use such a task.

Yes, I think we are saying the same thing here, there are always going to be cases which we couldn't handle automatically, but devs on those projects would probably want to add the dep manually anyway.

> I believe that AST-based solution would handle all edge-cases of custom templates and is exactly in the middle between "perfect" and simple code.

My experience working with AST so far has been in Credo on the reading side of things. My understanding is that it gets quite messy when you want to write back out modified AST in such a way that doesn't mangle original formatting or comments outside of the specific line you want to modify. This is why I approached the feature from the string manipulation angle, trading away some ability to identify the right change in certain approaches, for the benefit of not potentially mangling the rest of the file. If you have suggestions on good resources or examples of modifying files based on AST manipulation, I would love to tackle this from the AST side of things.

**Comment by Eiji7** [comment]
@TheFirstAvenger 

> if your deps function was multi-headed

You know what? I just got an idea that we could give a message similar to `FunctionClauseError`, but with a choice which clause (also multi and extra option for all) we should work on and if needed/possible a nice `CLI` way to do that automatically. When I think about it like it it may be worth to create an `archive` or `library` for such complex things as similar tasks may use it as well (for example task to auto-upgrade `Phoenix` from `1.2.x` to `1.3.x`).

> My understanding is that it gets quite messy when you want to write back out modified AST in such a way that doesn't mangle original formatting or comments outside of the specific line you want to modify.

Well … in case of formatting I would prefer to not complicate it and require an `Elixir` standard. Without standards everything goes messy sooner or later. However comments are really problematic thing … There are 2 ways to make it work:

1. Something around `AST` - for example adding it as a separate node (`{:#, [], ["comment"]}`) or as an meta option (`{:variable, [comment: "comment"], nil}`).
This way would give as a stable standard and allows us to traverse any `Elixir` code with or without comments. However it's definitely slower way. Also unfortunately it looks like a dead topic as previously in many places there was talk about such feature and their problems for example:
https://groups.google.com/g/elixir-lang-core/c/GM0yM5Su1Zc

2. Adding an extra option for API like `Code.format_string!/2` by passing a function which would traverse and modify quote `forms` right before formatting it (changing it to algebra - see `Code.Formatter.to_algebra/2`).
Also from `Elixir` core you may copy-paste important non-public code and provide an `API` like this. This way looks definitely faster. However it would require to maintain it and in worst case every new `Elixir` release you would need to compare your modified version with original code. Simply adding an extra option looks more reasonable, but not sure if that would happen soon.

I'm not happy with `Regex`/`String`-based solution, because I saw how those failed in practice. On the other side it's not trivial to preserve comments in traversed `AST` and that's why those topics are still not resolved. I would be looking forward if you decide to provide a nice `AST` traverse way as it would help in much more cases than such a simple task. Good luck! :crossed_fingers: :four_leaf_clover: :heart:

**Comment by bbugh** [comment]
@TheFirstAvenger wondering about why you decided to close this! Was there a status update that made it infeasible?

I'd really love to see this, coming from rust, javascript, ruby, python, etc. that allow you to add latest version of a dependency from the primary web source with a command line. Adding dependencies with elixir is a pain in comparison.
